### PR TITLE
Test retrieval of entities deleted from Datastore

### PR DESCRIPTION
**Pull request by anooshrees**
Tests the functionality added in PR #108 that handles servlets trying to retrieve Entities in Datastore that have been deleted and no longer exist.

### CODE REVIEW

File: src/test/java/com/google/sps/servlets/ContestServletTest.java [hunk 1]
```diff
@@ -327,9 +331,97 @@ public void oneElection_contestMissingInDatastore_testDoGet() throws IOException
...
+
+  @Test
+  public void oneElection_candidateMissingInDatastore_testDoGet() throws IOException {
+    Entity electionEntity = new Entity("Election");
+    Entity contestEntity = new Entity("Contest");
+
+    Entity firstCandidateEntity = new Entity("Candidate");
+    Entity secondCandidateEntity = new Entity("Candidate");
+
+    electionEntity.setPropertiesFrom(electionEntityOne);
```
Comment (julian-yang) [inline src/test/java/com/google/sps/servlets/ContestServletTest.java:353]: Can you group this with lines 347-348

File: src/test/java/com/google/sps/servlets/ContestServletTest.java [hunk 2]
```diff
@@ -327,9 +331,97 @@ public void oneElection_contestMissingInDatastore_testDoGet() throws IOException
...
+    firstCandidateEntity.setPropertiesFrom(candidateEntityOne);
+    secondCandidateEntity.setPropertiesFrom(candidateEntityTwo);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    long contestId = ds.put(contestEntity).getId();
+    long firstCandidateId = ds.put(firstCandidateEntity).getId();
+
+    // Delete the Entity from Datastore to mimic what happens at the beginning
+    // of a call to doGet() in ElectionServlet.
+    long secondCandidateId = ds.put(secondCandidateEntity).getId();
```
Comment (julian-yang) [inline src/test/java/com/google/sps/servlets/ContestServletTest.java:365]: I don't quite understand the point of this.  Aren't you just inserting something and then promptly deleting it?

File: src/test/java/com/google/sps/servlets/ContestServletTest.java [hunk 3]
```diff
@@ -327,9 +331,97 @@ public void oneElection_contestMissingInDatastore_testDoGet() throws IOException
...
+    Entity secondReferendumEntity = new Entity("Referendum");
+
+    electionEntity.setPropertiesFrom(electionEntityOne);
+    firstReferendumEntity.setPropertiesFrom(referendumEntityOne);
+    secondReferendumEntity.setPropertiesFrom(referendumEntityTwo);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    long referendumOneId = ds.put(firstReferendumEntity).getId();
+
+    long referendumTwoId = ds.put(secondReferendumEntity).getId();
```
Comment (julian-yang) [inline src/test/java/com/google/sps/servlets/ContestServletTest.java:405]: Couldn't you just make up some random ID instead?

File: src/test/java/com/google/sps/servlets/ContestServletTest.java [hunk 4]
```diff
@@ -327,9 +331,97 @@ public void oneElection_contestMissingInDatastore_testDoGet() throws IOException
...
+    verify(printWriter)
+        .println(
+            "{\"contests\":[{\"name\":\"myFirstContest\",\"candidates\":[{\"name\":\"myFirstCandidate\","
+                + "\"partyAffiliation\":\"myParty\",\"campaignSite\":\"myWebsite\",\"platformDescription\":"
+                + "\"This is a cool candidate.\",\"channels\":{}}],"
+                + "\"description\":\"This contest is important.\",\"source\":\"Voter Information Project\"}],\"referendums\":[]}");
+  }
+
+  @Test
+  public void oneElection_referendumMissingInDatastore_testDoGet() throws IOException {
```
Comment (julian-yang) [inline src/test/java/com/google/sps/servlets/ContestServletTest.java:393]: can you change this part to describe what the expected result is?
e.g. omits missing referendum or something like that

**Review by julian-yang — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/data/Contest.java [hunk 5]
```diff
@@ -134,7 +134,12 @@ public String toJsonString(DatastoreService datastore) {
                     : JsonNull.INSTANCE)
         // Have to use forEach to have void return. Can't use Collection to collect because
         // JsonArray can't addAll() with String as parameter.
-        .forEach(jsonElement -> candidateJsonArray.add(jsonElement));
+        .forEach(
```
[the remark this answers is not in the document]
Comment (rogerlaucorp) [reply src/main/java/com/google/sps/data/Contest.java:137]: +1

File: src/test/java/com/google/sps/servlets/ContestServletTest.java (hunk 2 above)
Comment (anooshrees) [reply src/test/java/com/google/sps/servlets/ContestServletTest.java:365]: This is to imitate what would happen when we do a mass deletion of entities when calling ElectionServlet/InfoCardServlet. Additionally, I can't get a non-zero ID value for the entities (a zero value ID will throw an error when I try to add it to the Contests entity) unless I add to Datastore.

File: src/test/java/com/google/sps/servlets/ContestServletTest.java (hunk 3 above)
Comment (anooshrees) [reply src/test/java/com/google/sps/servlets/ContestServletTest.java:405]: I could, but then I run the risk of generating a random ID that is already assigned to an Entity in Datastore. The deletion imitates what happens in the actual Servlet code.

File: src/test/java/com/google/sps/servlets/ContestServletTest.java (hunk 4 above)
Comment (anooshrees) [reply src/test/java/com/google/sps/servlets/ContestServletTest.java:393]: yes, will do!

**Comment by anooshrees** [comment]
Waiting for confirmation on approach from @julian-yang and then will push commits.

**Review by julian-yang — APPROVED** [review]

**Pull request merged**